### DISCUSSION TITLE
Partial Relationship Field

### DISCUSSION THREAD

**Discussion by 60pfennig** [payloadcms/payload#4896] [Feature Requests & Ideas]
Relationships are powerfull fields but they are adding tremendous overfetching. For example if you have a link like in the payload demo which relies on referencing to another page the data for every page will get fetched on the page which links to them. Even with maxDepth specified this effect leads to a multiple on API response size.
Another problem is that if you want to reference to for example posts in a post slider you just need a selection of fields which are important for the slider and the overview. You don't need the complete post. This seems to be neglectable but if you hava multiple of relationships and not just posts but bigger data structures this also leads to much bigger API reponses.

To tackle this I think it would be a good idea to have a mapable, partial relationship field. This could get archieved by a optional map attribute which takes a generic function which is called with every selected item of the referenced colletion and returns a type which has the "mandatory" field like id, but the rest is completly flexible. So i could return just the title an the image of a post or the path of the link instead of the complete collection item.

**Comment by WouterK12** [comment]
Hi, a solution to this was proposed in https://github.com/payloadcms/payload/discussions/2167#discussioncomment-5080370, but never implemented.

It would be a very helpful feature so hopefully it will be considered for a future update! :)

**Reply by WouterK12** [reply]
For now I ended up adding an [`afterRead`](https://payloadcms.com/docs/hooks/collections#afterread) hook to the collection.
The query that finds relationship data is specific enough that you can check if it's being executed, after which you can return a simplified version of the document:

```ts
import type { AfterReadHook } from 'payload/dist/collections/config/types';
import type { Where, WhereField } from 'payload/types';

type QueryType = Record<string, unknown> | undefined;

const isRelationshipQueryForDocId = (query: QueryType, docId: string) => {
    if (!query) {
        return false;
    }

    const where = query as Where;
    if (!where.and?.length) {
        return false;
    }

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (!where.and[0].id) {
        return false;
    }

    const field = where.and[0].id as WhereField;
    if (!field.in) {
        return false;
    }

    const ids = field.in as Array<string>;
    if (!ids.length) {
        return false;
    }

    return ids.includes(docId);
};

const afterReadHook: AfterReadHook<Page> = ({ doc, query }) => {
    const isRelationshipQuery = isRelationshipQueryForDocId(query, doc.id);

    if (isRelationshipQuery) {
        return {
            id: doc.id,
            slug: doc.slug,
        };
    }

    return doc;
};
```

This is not ideal, but while a proper solution isn't present this works fine :)
